components/brave_wallet_ui/common/slices/endpoints/nfts.endpoints.ts

```
@@ -0,0 +1,249 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import { EntityId } from '@reduxjs/toolkit'
+
+// types
+import {
+  BraveWallet,
+  ERC721Metadata,
+  NFTMetadataReturnType
+} from '../../../constants/types'
+import { WalletApiEndpointBuilderParams } from '../api-base.slice'
+
+// utils
+import {
+  GetBlockchainTokenIdArg,
+  getAssetIdKey
+} from '../../../utils/asset-utils'
+import { cacher } from '../../../utils/query-cache-utils'
+
+// entities
+import { blockchainTokenEntityAdaptor } from '../entities/blockchain-token.entity'
+
+// api
+import { getNetwork } from '../api.slice'
+import { translateToNftGateway } from '../../async/lib'
+
+export const nftsEndpoints = ({
+  query,
+  mutation
+}: WalletApiEndpointBuilderParams) => {
+  return {
+    getERC721Metadata: query<
+      {
+        id: EntityId
+        metadata?: ERC721Metadata
+      },
+      GetBlockchainTokenIdArg
+    >({
+      queryFn: async (tokenArg, _api, _extraOptions, baseQuery) => {
+        try {
+          if (!tokenArg.isErc721) {
+            return {
+              error: 'Cannot fetch erc-721 metadata for non erc-721 token'
+            }
+          }
+
+          const { jsonRpcService } = baseQuery(undefined).data
+
+          const result = await jsonRpcService.getERC721Metadata(
+            tokenArg.contractAddress,
+            tokenArg.tokenId,
+            tokenArg.chainId
+          )
+
+          if (result.error || result.errorMessage) {
+            return { error: result.errorMessage }
```

please throw this error